File: 2021/src/day07/golang/day07.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

func calcFuelForValue(guess int, crabPos []int, fuelCalc func(int) int, results chan<- int) {
	fuel := 0
	for _, pos := range crabPos {
		fuel += fuelCalc(Abs(guess - pos))
	}
	results <- fuel
}

func caclFuel(lines []string, fuelCalc func(int) int) int {
	crabPosStr := strings.Split(lines[0], ",")
	crabPos := make([]int, len(crabPosStr))
	for i, pos := range crabPosStr {
		crabPos[i], _ = strconv.Atoi(pos)
	}

	minValue := MinOf(crabPos...)
	maxValue := MaxOf(crabPos...)
	results := make(chan int, maxValue-minValue)
	for i := minValue; i < maxValue; i++ {
		go calcFuelForValue(i, crabPos, fuelCalc, results)
	}

	minFuel := math.MaxInt32
	for i := 0; i < maxValue-minValue; i++ {
		minFuel = MinOf(<-results, minFuel)
	}
	return minFuel
}

func part1(lines []string) int {
	return caclFuel(lines, func(nb int) int {
		return nb
	})
}

func part2(lines []string) int {
	return caclFuel(lines, func(nb int) int {
		return nb * (1 + nb) / 2
	})
}

func main() {

	if len(os.Args) < 2 {
		fmt.Println("Usage: go run {file.go} {filename}")
		return
	}

	file, err := os.Open(os.Args[1])
	if err != nil {
		fmt.Println("Unable to open file", err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	lines := make([]string, 0)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	fmt.Printf("part1 => %d\n", part1(lines))
	fmt.Printf("part2 => %d\n", part2(lines))
}

File: 2021/src/day07/Day07.kt
package day07

import readInput
import kotlin.math.abs

fun part1(lines: List<String>) : Int {
    val crabsPos = lines.first().split(",").map{ it.toInt() }
    val min = crabsPos.minOrNull() ?: return -1
    val max = crabsPos.maxOrNull() ?: return -1

    var minFuel = Integer.MAX_VALUE
    for (i in min..max) {
        var fuel = 0

        for (pos in crabsPos) {
            fuel += abs(i - pos)
        }
        if (fuel <= minFuel) minFuel = fuel
    }
    return minFuel
}

fun calcFuel(steps: Int, memo: MutableMap<Int, Int>) : Int {
    return if (steps <= 1) {
        steps
    } else {
        if (!memo.containsKey(steps)) {
            memo[steps] = steps + calcFuel(steps - 1, memo)
        }
        memo[steps]!!
    }
}

fun part2(lines: List<String>) : Int {
    val crabsPos = lines.first().split(",").map{ it.toInt() }
    val min = crabsPos.minOrNull() ?: return -1
    val max = crabsPos.maxOrNull() ?: return -1
    val memo = mutableMapOf<Int, Int>() // steps => fuel required map

    var minFuel = Integer.MAX_VALUE
    for (i in min..max) {
        var fuel = 0

        for (pos in crabsPos) {
            fuel += calcFuel(abs(i - pos), memo)
        }
        if (fuel <= minFuel) minFuel = fuel
    }
    return minFuel
}

fun main() {
    val testInput = readInput("day07/test")
    println("part1(testInput) => " + part1(testInput))
    println("part2(testInput) => " + part2(testInput))

    val input = readInput("day07/input")
    println("part1(input) => " + part1(input))
    println("part1(input) => " + part2(input))
}
